### PR TITLE
Secure Resolvers / LimitAccess Design Proposal

### PR DESCRIPTION
**Pull request by KirilKabakchiev**
**Description**

Changes proposed in this pull request:
- prepare a design proposal for limitAccess directive

### CODE REVIEW

**Comment by KirilKabakchiev** [comment]
Thanks for the feedback. I've added the proposal as well as another idea for the table that lands somewhere inbetween. Could you have another look ?

File: docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md [hunk 1]
```diff
@@ -1,39 +1,41 @@
...
+In general, integration systems can manage other systems and their metadata (such as applications, and runtimes). However, an integration system must be allowed to modify only entities that it has created, or entities to which it has granted access for modifications. For example, if an integration system registers an application, later on, it can  also register packages for this application. However, it cannot modify metadata for applications that other consumers registered, unless it has granted access explicitly to do so. Another example is when a user registers an application via the UI. To get access to this application, the integration system X either must be granted this access by the user via an API call to Director (grant access to the integration system for this application), or some automated procedures must be in place (a concept about stacking credentials via one time tokens is available in the further sections of this document). Although the given examples are related to applications, these requirements are also valid for application sub-entities, application templates, and runtimes managed by integration systems.
```
Comment (KirilKabakchiev) [inline docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md:22]: or entities to which it has granted access for modifications -> or entities to which it has *been* granted access for modification

Comment (KirilKabakchiev) [reply docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md:22]: unless it has granted access explicitly to do so ->  unless it has granted *been* access explicitly to do so

File: docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md [hunk 2]
```diff
@@ -1,39 +1,41 @@
...
+The problem with securing resolvers is three dimensional. It features all three: mutations/queries; consumer, and owner. In total, there are about 40 mutations/queries that have to be secured, 4 consumer types, and 4 owner types. For specific queries/mutations there is always a single owner but obtaining the id of that owner is not always straight forward. For example, in getApplication(applicationID) the owner id is passed as input to the query, however, in updatePackage(packageID) some special code needs to be executed to find the application id from the package id. In the case of updateAPIDefinition(apiDefID) the special code that needs to be executed to derive (nck: To derive what?).
```
Comment (KirilKabakchiev) [inline docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md:34]: In the case of updateAPIDefinition(apiDefID) the special code that needs to be executed to derive (nck: To derive what?). -> just remove it, the prev example is enough. You can put `updatePackage(packageID) or  updateAPIDefinition(apiDefID)` in the previous sentence

File: docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md [hunk 3]
```diff
@@ -1,39 +1,41 @@
...
+Why it is important who is the owner and what is the owner ID?
+
+Currently, with the help of the ORY integration, the Director API always recieves an ID token, regardless of the actual authentication mechanism. Then, the details about the actual caller/consumer can be extracted from the ID token, and then, stored in the Go context. Therefore, on a very high level, it is needed to ensure that the consumer from the Go context matches to the owner of the requested resource. More broadly put (and required especially for integration systems), it is important to verify that the consumer can access the owner's resources. For example, if the owner is an application, the resources that must be accessible are the application details, packages, and so on.
```
Comment (KirilKabakchiev) [inline docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md:38]: it  is needed to ensure -> it is necessary to ensure

Comment (KirilKabakchiev) [reply docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md:38]: matches to the owner -> matches the owner

File: docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md [hunk 4]
```diff
@@ -1,39 +1,41 @@
...
+This document describes a possible solution to the problem of unlimited resource access. The problem that needs to be corrected is the current possibility of one application in tenant X to modify the metadata of another application in the same tenant X. This problem is also valid for runtimes. Furthermore, integration systems have global access to all data in Compass. An optimal solution features an integration system that is limited to access only its own resources (newly created in the system) or resources, to which the system has granted access already.
 
-The proposal includes limiting access to operations for `applications`, `application_templates`, `runtimes` and `integration systems`.
+The solution proposal includes limiting the access to operations for `applications`, `application_templates`, `runtimes`, and `integration systems`.
```
Comment (KirilKabakchiev) [inline docs/internal/investigations/secure-resolvers/secure-resolvers-v2.md:5]: not sure if comma before and is needed?

**Review by nkanchev — APPROVED** [review]

**Review by desislavaa — APPROVED** [review]

**Review by dpanayotov — APPROVED** [review]

**Pull request merged**